Add redesigned connector implementation (#87)
ballerina-platform/module-salesforce: README.md

Connects to Salesforce from Ballerina. 

The Salesforce connector allows you to perform CRUD operations for SObjects, query using SOQL, search using SOSL, and

> describe SObjects and organizational data through the Salesforce REST API. Also it supports adding bulk data jobs and batches of types JSON, XML and CSV via Salesforce Bulk API. It handles OAuth 2.0 authentication.

Reviewer: ```suggestion
describe SObjects and organizational data through the Salesforce REST API. Also it supports adding bulk data jobs and batches of types JSON, XML and CSV via Salesforce Bulk API. It handles OAuth 2.0 authentication.
```
```suggestion
describe SObjects and organizational data through the Salesforce REST API. Also, it supports adding bulk data jobs and batches of types JSON, XML, and CSV via the Salesforce Bulk API. It handles OAuth 2.0 authentication.
```

**Base Client Operations**

> The base client of the Salesforce connector takes the Salesforce configurations in and creates other clients such as Sobject client, Query client and Bulk client. This also has some operations to retrieve available API versions and resources from Salesforce.

Reviewer: ```suggestion
The base client of the Salesforce connector takes the Salesforce configurations in and creates other clients such as the SObject client, Query client, and Bulk client. This also has some operations to retrieve available API versions and resources from Salesforce.
```

**SObject Operations**

The `ballerinax/sfdc48` module contains operations related to standard and customized SObjects. It can get, create, update, and delete SObjects via SObject IDs.

**SOQL & SOSL Operations**

> The `ballerinax/sfdc48` module contains operations that query using SOQL and search using SOSL. This allows for complex 

Reviewer: ```suggestion
The `ballerinax/sfdc48` module contains operations that query using SOQL and search using SOSL. This allows complex 
```

operations using SObjects relationships.

**Bulk Operations**

> The `ballerinax/sfdc48` module support bulk data operations for CSV, JSON and XML data types.

Reviewer: ```suggestion
The `ballerinax/sfdc48` module support bulk data operations for CSV, JSON, and XML data types.
```

   obtaining OAuth2 credentials, go to 
   [Salesforce documentation](https://help.salesforce.com/articleView?id=remoteaccess_authenticate_overview.htm).

> **Create Salesforce Base client**

Reviewer: ```suggestion
**Create the Salesforce Base client**
```

> You can define the Salesforce configuration and create Salesforce base client as mentioned below. 

Reviewer: ```suggestion
You can define the Salesforce configuration and create the Salesforce base client as mentioned below. 
```

};
```

> Using the base client, we can obtain the other clients.

Reviewer: ```suggestion
Using the base client, you can obtain the other clients.
```

```ballerina
sfdc48:SObjectClient sobjectClient = baseClient->getSobjectClient();

sfdc48:QueryClient queryClient = baseClient->getQueryClient();

sfdc48:BulkClient bulkClient = baseClient->getBulkClient();
```

**SObject Client Operations**

The `createRecord` remote function of the SObject client creates an SObject record. Pass a JSON object with the relevant fields needed for 
the SObject and the SObject type.

```ballerina
json account = { Name: "ABC Inc", BillingCity: "New York" };

string|sfdc48:ConnectorError createReponse = sobjectClient->createRecord("Account", account);
```

> The response from `createRecord` is either the string ID of the created record (if the record was created 

Reviewer: ```suggestion
The response from the `createRecord` is either the string ID of the created record (if the record was created 
```

}
```

**Query Client Operations**

The `getQueryResult` remote function of the Query client executes a SOQL query that returns all the results in a single response or if it 
exceeds the maximum record limit, it returns part of the results and an identifier that can be used to retrieve the 
remaining results.

```ballerina
string sampleQuery = "SELECT name FROM Account";

sfdc48:SoqlResult|sfdc48:ConnectorError response = queryClient->getQueryResult(sampleQuery);
```

The response from `getQueryResult` is either a SoqlResult record with total size, execution status, resulting records, 
and URL to get next record set (if query execution was successful) or `ConnectorError` (if the query execution 
was unsuccessful).

```ballerina
if (response is sfdc48:SoqlResult) {
    io:println("TotalSize:  ", response.totalSize.toString());
    io:println("Done:  ", response.done.toString());
    io:println("Records: ", response.records.toString());
} else {
    io:println("Error: ", response.detail()?.message.toString());
}
```

**Bulk Client Operations**

> Using the `createJob` remote function of the bulk client, we can create any type of job and of the data type JSON, XML and CSV. 

Reviewer: ```suggestion
Using the `createJob` remote function of the bulk client, you can create any type of job of the data type JSON, XML, and CSV. 
```

```ballerina
error|sfdc48:BulkJob insertJob = bulkClient->creatJob("insert", "Contact", "JSON");
```

> Using the created job object, we can add batch to it, get information about the batch and get all the batches of the job.

Reviewer: ```suggestion
Using the created job object, you can add a batch to it, get information about the batch, and get all the batches of the job.
```

```ballerina
    json contacts = [
        {
            description: "Created_from_Ballerina_Sf_Bulk_API",
            FirstName: "Morne",
            LastName: "Morkel",
            Title: "Professor Grade 03",
            Phone: "0442226670",
            Email: "morne89@gmail.com"
        }
    ];

>     //add json content

Reviewer: ```suggestion
    //Add the JSON content.
```

    error|sfdc48:BatchInfo batch = insertJob->addBatch(contacts); 
```   

```

>     //get batch info

Reviewer: ```suggestion
    //Get batch info.
```

    error|sfdc48:BatchInfo batchInfo = insertJob->getBatchInfo(batch.id
``` 

```

>     //get all batches

Reviewer: ```suggestion
    //Get all batches.
```

    error|sfdc48:BatchInfo[] batchInfoList = insertJob->getAllBatches();
```

```

>     //get batch request

Reviewer: ```suggestion
    //Get the batch request.
```

    var batchRequest = insertJob->getBatchRequest(batchId);
```

```

>     //get batch result

Reviewer: ```suggestion
    //Get the batch result.
```

    error|sdfc48:Result[] batchResult = insertJob->getBatchResult(batchId);
```

> `getJobInfo` remote function of the bulk client retrieves all details for an existing job.

Reviewer: ```suggestion
The `getJobInfo` remote function of the bulk client retrieves all details of an existing job.
```

```ballerina
    error|sfdc48:JobInfo jobInfo = bulkClient->getJobInfo(insertJob);
```

> `closeJob` and `abortJob` remote functions close and abort CSV insert job respectively. When a job is closed, no more 

Reviewer: ```suggestion
The `closeJob` and `abortJob` remote functions close and abort the CSV insert job respectively. When a job is closed, no more 
```
